File: NDK_Day26_AS/app/src/main/java/com/darren/ndk/day26/MainActivity.java
package com.darren.ndk.day26;

import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;

public class MainActivity extends AppCompatActivity {

    // Used to load the 'native-lib' library on application startup.
    static {
        System.loadLibrary("native-lib");
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        Person[] persons1 = new Person[10];
        for (int i = 0; i < persons1.length; i++) {
            persons1[i] = new Person("Darren", 24 + i);
        }

        // java 层赋值
        Person[] newPersons = new Person[10];
        /*
        for (int i = 0; i < newPersons.length; i++) {
            newPersons[i] = persons1[i];
        }*/

        try {
            arraycopy(persons1, 0, newPersons, 0, 10);
        } catch (Exception e) {
            e.printStackTrace();
        }


        for (int i = 0; i < newPersons.length; i++) {
            Log.e("TAG", newPersons[i].toString());
        }
    }

    public static native void arraycopy(Object src, int srcPos,
                                        Object dest, int destPos,
                                        int length);
}

File: NDK_Day14/app/src/main/java/com/darren/ndk/day14/MainActivity.kt
package com.darren.ndk.day14

import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.util.Log
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        // Example of a call to a native method
        sample_text.text = stringFromJNI()
        var parcel = Parcel();

        parcel.writeInt(12)
        parcel.writeInt(24)

        parcel.setDataPosition(0)

        var number1 = parcel.readInt()
        var number2 = parcel.readInt()

        Log.e("TAG","number1 = $number1 , number2 = $number2")
    }

    /**
     * A native method that is implemented by the 'native-lib' native library,
     * which is packaged with this application.
     */
    external fun stringFromJNI(): String

    companion object {

        // Used to load the 'native-lib' library on application startup.
        init {
            System.loadLibrary("native-lib")
        }
    }
}
